### ISSUE TITLE
overly lenient input type for take, drop, split-at, take-right, drop-right, split-at-right

### ISSUE THREAD

**Issue by mbutterick** [racket/typed-racket#67]
In TR, all these functions take an `Integer` argument. In Racket, however, these arguments are limited to `exact-nonnegative-integer?`. So the TR types can be tightened to either `Nonnegative-Integer` or I assume `Index` (since they pertain to list indexes).

**Comment by mbutterick** [comment]
(Clarification: take an `Integer` argument for the list index where the action should occur)

**Comment by samth** [comment]
We experimented with this approach for a while; there's actually still an abstraction in TR for switching this. But we found that it was quite inconvenient, especially because subtraction doesn't produce an index. So we switched to the types you see now.

**Comment by mbutterick** [comment]
OK. More broadly, when the index argument in these functions is bound at compile time, it has immediate implications for types, so I was curious about how I might leverage that fact. For instance `take`:

```racket
(All (a) (-> (Listof a) Integer (Listof a)))
```

Let’s suppose every integer datum were its own type, just as every symbol is its own type. (This may not be something that's feasible or interesting to you, but hypos are cheap.)

Then the parameterization could conceivably not be just by type, but by value:

```racket
(All (a x) (-> (Listof a) x (Listof a)))
```
 
Moreover, suppose we're doing `(take xs 3)`. This tells us a) that `xs` needs to have length >= `3`, and b) our return value will have exactly `3` elements. So the type could be refined:

```racket
(All (a x) (-> (List* a a (Listof a)) x (List a a a)))
```

Then the question would be how to make this typing parametric relative to `x`, for instance with type constructors that can read the value:

```racket
(All (a x) (-> (List-Min-Length a x) x (List-Exact-Length a x)))
```

Though perhaps the penultimate step is just a macro-expansion away from the last.

**Comment by rfindler** [comment]
Dependent types, whoo hoo!

**Comment by andmkent** [comment]
Matthew - I'm currently in the process of adding some basic dependent types
to Typed Racket (refinement types that can restrict a type with logical
propositions over run-time types and integer constraints on linear
combinations), for which I would also like to express 'whoo hoo!' =)

This particular case you mention seems like just the sort of thing would
could describe using this extension since it would allow you to refine the
length of the output list based on an input value. (It also adds types that
describe each integer datum, as you suggested).

By chance, are you thinking of some already existing code that you would
like to use these sort of types on? If so, and if you wouldn't mind
sharing, I would love to take a look at it. Real world programs that
require your feature are always nice to know about! =)

Best,
Andrew

**Comment by mbutterick** [comment]
Yes. I’m working on a TeX-style typesetting engine called [Quad](http://github.com/mbutterick/quad). It’s on the package server too, and it’s starting to work, though I haven’t set up any simple demos (I should). 

I prototyped it in standard Racket. It was too slow, so I’m converting it to TR. I made a first pass just to get everything under `#lang typed/racket/base`. On the second pass, I’m starting to get the jiu-jitsu of typing everything as tightly as possible. 

The core data structure in Quad is the eponymous Quad: like an X-expression, it's a list with a name, sublist of attributes, and sublist of elements, so the type is roughly: 

`(List Symbol (Listof (Pairof Symbol Any)) (Listof (U String Quad)))`

Most of the program, therefore, is list-juggling functions. For instance, the function that raised this particular question splits the last element off a list:

```racket
(define/typed+provide (split-last xs)
  (All (A) ((Listof A) -> (values (Listof A) A)))
  (let-values ([(first-list last-list) ((inst split-at-right A) xs 1)])
    (values first-list (car last-list))))
```

Since performance counts, one goal is to put functions into a form that TR can optimize with a minimum of runtime checks.  (See issue #61 for another example.) In this case, there’s enough compile-time information (namely, the `1` datum) to optimize the `(car last-list)` in the last line, but there’s currently no typing abstraction available to connect them.

So yes, I’m sure I could find a use for your dependent types.

**Comment by rfindler** [comment]
There may be more efficient representations than xexprs (that could support
similar notations and apis), depending on the juggling you need to do.

Robby


On Sat, Mar 28, 2015 at 1:00 PM, Matthew Butterick <notifications@github.com

**Comment by mbutterick** [comment]
Suggestions welcome. I tried combinations of structs, hash tables, and vectors, but the simpler list structure was the best fit so far with TR typing and optimization.
